nova_vm/src/ecmascript/builtins/indexed_collections/array_objects/array_prototype.rs

```diff
@@ -3101,6 +3101,35 @@ impl ArrayPrototype {
         this_value: Value,
         items: ArgumentsList,
     ) -> JsResult<Value> {
+        // Fast path: Array is dense and contains no descriptors. No JS
+        // functions can thus be called by unshift.
+        if let Value::Array(array) = this_value {
+            if array.is_trivial(agent)
+                && array.is_dense(agent)
+                && agent[array].elements.len_writable
```

Comment:
**issue:** Before going into this branch, we should also check that `len + arg_count <= u32::MAX`. The `some_usize as u32` doesn't saturate (clamp itself to the maximum value) but instead wraps around (ignores highest bits).

Right now, trying to unshift more data into the Array than can be indexed using u32 will cause a panic.

Reply:
Fixed by https://github.com/trynova/nova/pull/461/commits/ee619df0d1365259f4cd97a444d652c96aa23894

Reply:
**issue:** Instead of that, do this:
```rs
if let Value::Array(array) = this_value {
  let len = array.len(agent);
  let arg_count = items.len();
  let final_len = u32::try_from(len + arg_count);
  if final_len.is_some() && array.is_trivial(agent) && array.is_dense(agent) && array.length_writable(agent) {
   // ...
```
(The last `length_writable` would be a worthwhile addition to the Array's API :) )

We need to make sure that the final length can be expressed as a `u32`; you only checked that it is within `u53`. And of course, we don't want to throw an error if we're eg. 1 over `u32::MAX`, that wouldn't be correct according to the spec. We'll just take the slow path in that case.